File: L520/520.cpp
#include <cctype>
using namespace std;

class Solution {
public:
    int cnt = 0;
    bool ans = false;

    void check_capital(string word, int start, int end) {
        int mid = start + (end-start)/2;
        if (start==end) {
            if (isupper(word[start])) cnt++;
        } else {
            check_capital(word, start, mid);
            check_capital(word, mid+1, end);
        }
    }

    bool detectCapitalUse(string word) {
        check_capital(word, 0, word.length()-1);
        if (word.length() == cnt) {
            ans = true;
        } else if (cnt == 0) {
            ans = true;
        } else if (cnt == 1 && isupper(word[0])) {
            ans = true;
        } else {
            ans = false;
        }
        return ans;
    }
};

File: L520/520.py
class Solution:
    def __init__(self):
        self.cnt = 0
        self.ans = False

    def capital_check(self, word: str) -> None:
        length = len(word)
        idx = length // 2
        if length == 1:
            self.cnt += word.isupper()
        else:
            self.capital_check(word[:idx])
            self.capital_check(word[idx:])

    def detectCapitalUse(self, word: str) -> bool:
        self.capital_check(word)
        if len(word) == self.cnt:
            self.ans = True
        elif self.cnt == 0:
            self.ans = True
        elif self.cnt == 1 & word[0].isupper():
            self.ans = True
        else:
            self.ans = False
        return self.ans
